Mergin/version_viewer_dialog.py

```diff
@@ -0,0 +1,588 @@
+        add_all_action = menu.addAction(QIcon(icon_path("folder-plus.svg")), "Add all changes layers to project")
+        add_all_action.triggered.connect(self.add_all_to_project)
+        btn_add_changes.setMenu(menu)
+
+        self.toolbar.addWidget(btn_add_changes)
+        self.toolbar.setIconSize(iface.iconSize())
+
+        self.map_canvas.enableAntiAliasing(True)
+        self.map_canvas.setSelectionColor(QColor(Qt.cyan))
+        self.pan_tool = QgsMapToolPan(self.map_canvas)
+        self.map_canvas.setMapTool(self.pan_tool)
+
+        self.current_diff = None
+        self.diff_layers = []
+        self.filter_model = None
+        self.layer_list.currentRowChanged.connect(self.diff_layer_changed)
+
+        self.icons = {
+            "added": "plus.svg",
+            "removed": "trash.svg",
+            "updated": "pencil.svg",
+            "renamed": "pencil.svg",
+            "table": "table.svg",
+        }
+        self.model_detail = QStandardItemModel()
+        self.model_detail.setHorizontalHeaderLabels(["Details"])
+        self.details_treeview.setModel(self.model_detail)
+
+        self.versionModel.current_version = self.mp.version()
+
+    def exec(self):
+        try:
+            ws_id = self.mp.workspace_id()
+        except ClientError as e:
+            QMessageBox.warning(None, "Client Error", str(e))
+            return
+
+        # check if user has permissions
+        usage = self.mc.workspace_usage(ws_id)
+        if not usage["view_history"]["allowed"]:
+            QMessageBox.warning(None, "Permission Error", "The workspace does not allow to view project history.")
+            return
+
+        self.reject()
+        return
+
+    def closeEvent(self, event):
+        self.save_splitters_state()
+        QDialog.closeEvent(self, event)
+
+    def save_splitters_state(self):
+        settings = QSettings()
+        settings.setValue("Mergin/VersionViewerSplitterSize", self.splitter.saveState())
+        settings.setValue("Mergin/VersionViewerSplitterVericalSize", self.splitter_vertical.saveState())
+
+    def set_splitters_state(self):
+        settings = QSettings()
+        state_vertical = settings.value("Mergin/VersionViewerSplitterVericalSize")
+        if state_vertical:
+            self.splitter_vertical.restoreState(state_vertical)
+        else:
+            self.splitter_vertical.setSizes([120, 200, 40])
+
+        state = settings.value("Mergin/VersionViewerSplitterSize")
+        if state:
+            self.splitter.restoreState(state)
+        else:
+            height = max([self.map_canvas.minimumSizeHint().height(), self.attribute_table.minimumSizeHint().height()])
+            self.splitter.setSizes([height, height])
+
+    def fetch_from_server(self):
+
+        if self.fetcher and self.fetcher.isRunning():
+            # Only fetching when previous is finshed
+            return
+
+        self.fetcher = VersionsFetcher(self.mc, self.mp.project_full_name(), self.versionModel)
```

Comment:
If you want to pass in the `versionModel` to the fetcher, I'd expect it would add the versions to the model directly, if not, do we even need to pass it in?